### ISSUE TITLE
[Feature Request] Keep only one module to handle path

### ISSUE THREAD

**Issue by gryznar** [modularml/mojo#1874]
### Review Mojo's priorities

- [X] I have read the [roadmap and priorities](https://docs.modular.com/mojo/roadmap.html#overall-priorities) and I believe this request falls within the priorities.

### What is your request?

In Mojo there are currently 2 modules to handle paths: `os.path` and `pathlib`. I am requesting to just replace `pathlib` -> `path` and merge content from `os.path`

### What is your motivation for this change?

I am aware that this approach tries to replicate Python, but this does not make to much sense in terms of names consistency, simplicity and maintenance. Instead of just copying Python, we may do it better and have only one object-oriented approach. Structs in Mojo is lightweight, so it shouldn't hit performance. `path` module will be still similar to Python, but won't introduce this need to adjust implementation to align with Python release

**Comment by gabrieldemarmiesse** [comment]
Doing so would require an entier design proposal, so a lot of bikeshedding and that would slow down Mojo's development. We'd also confuse Python users. 

The pathlib [PEP](https://peps.python.org/pep-0428/) makes it clear that pathlib is for high level operation and os.path for low level operations. Thus not exposing complexity to most users as they'll only interact with pathlib. This is mentioned in the python documentation of both modules.

If you believe that this is the wrong approach to adopt and that only one module should be used for both low-level and high-level operations, then feel free to make a PEP, I'm sure the Mojo team will be happy to implement it when it's accepted by Python's steering council. They explain it in the roadmap:
> If/when Python adopts a feature, Mojo will also add it, because Mojo's goal is to be a superset. We are happy with this approach because the Python community is better equipped to evaluate these features, they have mature code bases to evaluate them with, and they have processes and infrastructure for making structured language evolution features.

**Comment by gryznar** [comment]
So, the question still arises. Should Mojo replicate Python stdlib? IMHO - no and there is a lot of reasons why not to do that (stdlib size, maintanance, consistency (naming in Python is a mess) and so on). It will be better to implement Python stdlib as an external library and allow people who wants compatibility with Python to just download it. This can be then automatically auto imported to allow direct imports

**Comment by gabrieldemarmiesse** [comment]
I'm not sure the Modular team wants to maintains two different stdlib. Futhermore it's going to split the community. We'll have to explain to everyone which stdlib to use. Some people like the python stdlib so they'll keep using it. We'll end up exactly in the situation that you are critisizing with os.path and pathlib, where we have two tools that can do the job and users don't know which one to use.

**Comment by gryznar** [comment]
Modular could just create its own stdlib (lightweiht and consistent with only essential stuff). Python compatibility layer may be added later via Modular itself or via community

**Comment by gabrieldemarmiesse** [comment]
I don't believe this aligns well with the goals of Mojo in https://docs.modular.com/mojo/why-mojo#a-member-of-the-python-family . Not sure the idea of re-inventing the stdlib will get much traction.

**Comment by gryznar** [comment]
This approach won't be compatible with accessing also devices with restricted resources where size matters. Mojo is aiming much further than just to be compiled Python

**Comment by soraros** [comment]
We can at best create a library that superficially resembles the Python standard library. This is due to, for instance, (a) value versus reference semantics and (b) traits versus subclassing. In the case of the former, it would be problematic if the same code silently produced different (and therefore incorrect) results. Regarding the latter, see, for example, issue #1744 (idiomatic Python patterns may prove ineffective in Mojo). Moreover, the infrastructure (dynamic Mojo, classes) necessary for writing code that accurately matches Python semantics is not yet available.

Despite sharing syntax, Mojo is still more akin to Rust or Swift than to Python. Consequently, the structure of its standard library must be different, making the reimplementing or redesigning of stdlib inevitable.
